keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/ArtifactUtils.kt

```diff
@@ -19,65 +19,36 @@ package com.netflix.spinnaker.keel.sql
 import com.fasterxml.jackson.databind.ObjectMapper
-import com.fasterxml.jackson.module.kotlin.convertValue
 import com.fasterxml.jackson.module.kotlin.readValue
 import com.netflix.spinnaker.exceptions.ArtifactParsingException
-import com.netflix.spinnaker.keel.api.artifacts.ArtifactStatus
 import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
-import com.netflix.spinnaker.keel.api.artifacts.ArtifactType.deb
-import com.netflix.spinnaker.keel.api.artifacts.ArtifactType.docker
 import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
-import com.netflix.spinnaker.keel.api.artifacts.VirtualMachineOptions
-import com.netflix.spinnaker.keel.artifacts.DebianArtifact
-import com.netflix.spinnaker.keel.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.api.plugins.ArtifactPublisher
 import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
 
 private val objectMapper: ObjectMapper = configuredObjectMapper()
 
 /**
- * A helper function to construct the proper artifact type from the serialized json.
- * FIXME: this needs to go away in favor of [ArtifactPublisher] functions to keep
- * artifact contracts generic.
+ * A helper function to construct the proper artifact type from the serialized JSON.
  */
 fun mapToArtifact(
+  artifactPublisher: ArtifactPublisher<*>,
   name: String,
   type: ArtifactType,
   json: String,
   reference: String,
   deliveryConfigName: String
 ): DeliveryArtifact {
   try {
-    val details = objectMapper.readValue<Map<String, Any>>(json)
-    return when (type) {
-      deb -> {
-        val statuses: Set<ArtifactStatus> = details["statuses"]?.let { it ->
-          try {
-            objectMapper.convertValue<Set<ArtifactStatus>>(it)
-          } catch (e: IllegalArgumentException) {
-            null
-          }
-        } ?: emptySet()
-        DebianArtifact(
-          name = name,
-          statuses = statuses,
-          reference = reference,
-          deliveryConfigName = deliveryConfigName,
-          vmOptions = details["vmOptions"]?.let {
-            objectMapper.convertValue<VirtualMachineOptions>(it)
-          } ?: error("vmOptions is required")
-        )
+    val artifactAsMap = objectMapper.readValue<Map<String, Any>>(json)
+      .toMutableMap()
+      .also {
+        it["name"] = name
+        it["type"] = type
+        it["reference"] = reference
+        it["deliveryConfigName"] = deliveryConfigName
       }
-      docker -> {
-        val tagVersionStrategy = details.getOrDefault("tagVersionStrategy", "semver-tag")
-        DockerArtifact(
-          name = name,
-          tagVersionStrategy = objectMapper.convertValue(tagVersionStrategy),
-          captureGroupRegex = details["captureGroupRegex"]?.toString(),
-          reference = reference,
-          deliveryConfigName = deliveryConfigName
-        )
-      }
-    }
+    return objectMapper.convertValue(artifactAsMap, artifactPublisher.supportedArtifact.artifactClass)
```

Comment:
Does this correctly convert a `null` status to an empty list for docker artifacts?

Reply:
It won't be null, it just won't be present in the JSON because of the `@JsonIgnore`.

Example for Debian:
```json
{"statuses": ["SNAPSHOT"], "vmOptions": {"baseOs": "xenial", "regions": ["us-west-2"], "baseLabel": "RELEASE", "storeType": "EBS"}}
```

Example for Docker:
```json
{"tagVersionStrategy": "branch-job-commit-by-job"}
```